src/pjpersist/tests/test_datamanager.py

```diff
@@ -1609,6 +1609,269 @@ def cursor(self, abort=False):
             transaction.commit()
 
 
+class DatamanagerFailedTransactionTest(testing.PJTestCase):
+
+    def setUp(self):
+        super(DatamanagerFailedTransactionTest, self).setUp()
+
+        # get rid of the previous transaction
+        transaction.abort()
+
+        tpc_patch = mock.patch(
+            "pjpersist.datamanager.PJ_TWO_PHASE_COMMIT_ENABLED", True)
+        self.patches = [tpc_patch]
+        for p in self.patches:
+            p.start()
+
+        # First PJDataManager instantiation creates tables, what makes the dm
+        # dirty, which we want to avoid here.
+        self.conn = testing.getConnection(testing.DBNAME)
+        self.dm = datamanager.PJDataManager(self.conn)
+
+    def tearDown(self):
+        for p in self.patches:
+            p.stop()
+
+        super(DatamanagerFailedTransactionTest, self).tearDown()
+
+    def beginTransaction(self, tpc=True):
+        datamanager.PJ_TWO_PHASE_COMMIT_ENABLED = tpc
+        transaction.abort()
+        self.dm = datamanager.PJDataManager(self.conn)
+
+    def test_broken_transaction_commit(self):
+        self.beginTransaction(tpc=False)
+
+        # Make changes
+        self.dm.root['foo1'] = Foo('foo-first')
+        # Someone broke the postgres transaction using dm's connection
+        with self.conn.cursor() as cursor:
+            with self.assertRaises(psycopg2.Error):
+                cursor.execute("FAULTY SQL")
+        # No more changes can be made
+        with self.assertRaises(psycopg2.errors.InFailedSqlTransaction):
```

Comment:
what happens if there's no "pj data change" after "FAULTY SQL" and before transaction.commit()?

I'd say transaction.commit() should fail then